### PR TITLE
API-17322/  loast  implement swagger parser for loast validation test

### PR DESCRIPTION
**Pull request by mheusgit**
- Added Spectral for linting OAS and included Spectral test suite validator
- Refactored validation, conductors, messaging, and commands to remove tight coupling between them and 'positive' validation testing.  Adopting a new 'test suite' approach with corresponding folders for relevant suites.
- Added Loast.ts to drive runs against multiple test suites per OAS
- Changed naming of some parameters and methods to avoid confusion with new test suites
- Updated TypeScript due to Spectral needing more recent functionality.  This change also needed oclif to updated to work with TypeScript version 
- Updated Jest to improve ESM module support and avoid resolution errors connected to Spectral
- Disabled eslint rules conflicting with prettier

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -23,14 +23,15 @@ module.exports = {
     semi: ["error", "always"],
     "@typescript-eslint/camelcase": "off",
     "comma-dangle": ["error", "always-multiline"],
-    indent: ["error", 2],
+    indent: ["off"],
```
Comment (stevelong00) [inline .eslintrc.js:26]: This `indent` rule can be omitted since the `"prettier/prettier"` rule on the next line does its own indentation checks.

Comment (mheusgit) [reply .eslintrc.js:26]: Omitting the indent rule is not enough.  I ran into scenarios where Prettier and eslint started conflicting over how the page should appear.

A good example was where I added the switch() statement for the SuiteFactory.  Prettier would make its changes and eslint would flag it as an error.

File: .eslintrc.js [hunk 2]
```diff
@@ -23,14 +23,15 @@ module.exports = {
...
     "prettier/prettier": ["error"],
     'lines-between-class-members': [
       'off'
     ],
     'max-params': [
       'off'
-    ]
+    ],
+    'unicorn/no-abusive-eslint-disable': [ 'off' ],
```
Comment (christineyost) [inline .eslintrc.js:34]: Why were these changes needed?

File: test/suites/positive/positive-suite.test.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+let suiteConfig: SuiteConfig;
+let oasSchema: OASSchema;
+operationResult.set(
+  '/findTheRing:GET',
+  new Map().set('no-rings-found', {
+    failures: { test: 'mockError' },
+    warnings: {},
+  }),
+);
```
Comment (christineyost) [inline test/suites/positive/positive-suite.test.ts:28]: It looks like this operationResult is similar or the same as the one used in `spectral-suite.test.ts`. Can this be defined as a fixture and reused in both these tests?

File: test/suites/suite-factory.test.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    it('throws error for unknown suites', () => {
+      const options = { path: 'fake-path', apikey: 'fake-key' };
+
+      expect(() =>
+        SuiteFactory.build('gibberishSuiteId', {
+          ...suiteConfig,
+          options: options,
+        }),
+      ).toThrowError();
```
Comment (christineyost) [inline test/suites/suite-factory.test.ts:55]: Can this test confirm an error is thrown with the message `Unable to find suite with ID ${suiteId}`?

**Review by christineyost — COMMENTED** [review]
Very exciting to get Spectral in LOAST!

I think we should avoid using `eslint-disable` where possible. I also think we could look into adding types for Spectral, like we did for `swagger-client` (https://github.com/department-of-veterans-affairs/lighthouse-oas-tests/blob/master/%40types/swagger-client/index.d.ts), but I think we could look into that separately

File: src/commands/suites-batch.ts [hunk 5]
```diff
@@ -1,8 +1,10 @@
...
-import { PositiveBatchConductor } from '../conductors';
-import { OASResult } from '../validation/results';
+import { OASResult } from '../validation';
+import { Config } from '../config';
+import { FileIn } from '../utilities/file-in';
+import Loast from '../loast';
 
-export default class PositiveBatch extends Command {
+export default class SuitesBatch extends Command {
   static description =
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/commands/suites-batch.ts:8]: Yep.  That needs a correction

File: src/suites/suite.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import { OperationResult } from '../validation';
+import { TestOptions } from '../config/config.interface';
+import { SecurityValues } from 'swagger-client';
+import OASSchema from '../oas-parsing/schema';
+
+export interface SuiteConfig {
+  options: TestOptions;
+  schema: OASSchema;
+  securityValues?: SecurityValues;
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/suite.ts:9]: No.  Other suites will probably not use it.  I was thinking it would be available if needed.

File: .eslintrc.js (hunk 2 above)
Comment (mheusgit) [reply .eslintrc.js:34]: [eslint ](https://github.com/department-of-veterans-affairs/lighthouse-oas-tests/pull/88#discussion_r957847993)

File: test/suites/positive/positive-suite.test.ts (hunk 3 above)
Comment (mheusgit) [reply test/suites/positive/positive-suite.test.ts:28]: Sure

**Review by dsteeve — COMMENTED** [review]
would like to talk thru the structure of suites and test groupings before going further.

File: .eslintrc.js (hunk 1 above)
Comment (mheusgit) [reply .eslintrc.js:26]: Attached two files showing eslint and prettier conflicting over indents.
Eslint complaining
prettier

Prettier complaining
eslint

I recommend to give prettier control over behavior.

File: src/loast.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    let suiteIds = this.options.loastType;
+
+    if (!suiteIds || suiteIds.length === 0) {
+      suiteIds = SuiteFactory.availableSuiteIds();
+    }
+
+    for (let x = 0; x < suiteIds.length; x++) {
+      // eslint-disable-next-line no-await-in-loop
+      results.push(await this.getSuiteResults(suiteIds[x]));
+    }
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/loast.ts:36]: For this PR I'll switch to the map approach

File: src/suites/positive/positive-suite.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+export default class PositiveSuite extends Suite {
+  public static suiteId = 'positive';
+  public static label = '(Example Group: Positive)';
+
+  async conduct(): Promise<OperationResult[]> {
+    this.checkTargetServer();
+
+    if (!this.suiteConfig.securityValues) {
+      throw new Error('Unable to run suite due to missing securityValues');
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/positive/positive-suite.ts:14]: The ValidationConductor expects it to be defined.  This was added to resolve TS errors.

File: src/suites/positive/validation/positive-validator.ts [hunk 9]
```diff
@@ -33,18 +34,6 @@ abstract class BaseValidator {
     }
   }
 
-  abstract performValidation(): void;
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/positive/validation/positive-validator.ts:36]: In the long term they should fully match but this ticket has to a stopping point somewhere.  I went with few changes for it.

File: src/suites/positive/positive-suite.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        throw new Error(
+          'Server value must match one of the server URLs in the OAS',
+        );
+      }
+    }
+  }
+
+  public getLabel(): string {
+    return PositiveSuite.label;
+  }
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/positive/positive-suite.ts:59]: The variable will be moved to 'protected'

File: src/suites/positive/validation/positive-validator.ts [hunk 11]
```diff
@@ -1,28 +1,29 @@
...
   }
 
-  public get warnings(): Map<string, ValidationMessage> {
+  public get warnings(): Map<string, Message> {
     return this._warnings;
   }
 
-  protected addMessage(type: Type, path: string[], props?: string[]): void {
-    const message = new ValidationMessage(type, path, props);
+  public addMessage(type: unknown, path: string[], props?: string[]): void {
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/positive/validation/positive-validator.ts:25]: This is a ongoing issue I've observed before with the current approach used for Jest testing since many methods are assigned as public then they should be private/protected in order to better support unit testing.

In this case it was also done to match the other validator.

File: src/suites/spectral/spectral-suite.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import Suite, { SuiteConfig } from '../suite';
+import SpectralValidator from './validation/spectral-validator';
+
+export default class SpectralSuite extends Suite {
+  public static suiteId = 'spectral';
+  public static label = '(Spectral)';
+
+  constructor(suiteConfig: SuiteConfig) {
+    super(suiteConfig);
+    this.suiteConfig = suiteConfig;
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/spectral/spectral-suite.ts:11]: No.  It used to have other logic afterwards but that was cleaned up.

File: src/suites/spectral/spectral-suite.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        }
+      }
+    }
+
+    return results;
+  }
+
+  public getLabel(): string {
+    return SpectralSuite.label;
+  }
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/spectral/spectral-suite.ts:52]: Not sure what the concern is here.

File: src/suites/spectral/validation/spectral-message.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    details: '{0}',
+  },
+  [Type.SpectralError]: {
+    severity: Severity.ERROR,
+    details: '{0}',
+  },
+};
+
+class SpectralMessage extends Message {
+  constructor(type: unknown, path: string[], _properties?: string[]) {
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/spectral/validation/spectral-message.ts:21]: it can.  I updated the positive-message.ts as well

File: src/suites/suite.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+  // conduct(): Expected to perform a series of related test case validations against an OAS
+  //  and return an array of results.
+  //  Should throw an error if testing cannot be performed due to a major issue (such as missing/invalid config)
+  abstract conduct(): Promise<OperationResult[]>;
+
+  // Should be overriden in child suite
+  public getLabel(): string {
+    return Suite.label;
+  }
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/suite.ts:30]: Variable was moved to 'protected'

File: src/suites/suite-factory.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        return new PositiveSuite(config);
+      case SpectralSuite.suiteId:
+        return new SpectralSuite(config);
+      default:
+        throw new Error(`Unable to find suite with ID ${suiteId}`);
+    }
+  }
+
+  public static availableSuiteIds(): string[] {
+    return [PositiveSuite.suiteId, SpectralSuite.suiteId];
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/suite-factory.ts:19]: The approach is used to map the classes to the requested loastType.  So such a change requires more than just a enumeration.

File: test/suites/positive/positive-suite.test.ts (hunk 3 above)
Comment (mheusgit) [reply test/suites/positive/positive-suite.test.ts:28]: From looking over the code I thought this was larger code-wise than what was shown.  It doesn't make sense to move this to create a fixture for 7 lines.

File: test/suites/suite-factory.test.ts (hunk 4 above)
Comment (mheusgit) [reply test/suites/suite-factory.test.ts:55]: Updated

File: src/commands/suites.ts [hunk 17]
```diff
@@ -38,43 +39,53 @@ export default class Positive extends Command {
   ];
 
   async run(): Promise<void> {
-    const { args, flags } = this.parse(Positive);
+    const { args, flags } = this.parse(Suites);
 
-    const positiveConductor = new PositiveConductor(DEFAULT_TEST_NAME, {
+    const loast = new Loast(DEFAULT_TEST_NAME, {
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/commands/suites.ts:44]: The different suites can be run by providing the 'loastType' as part of the TestOptions.  This gives full control over what test are run for both the suites & suite-batch commands.

If no loastType is defined, all registered test suites in SuiteFactory are used.
If loastType = ['spectral'] then only the Spectral suite will be run.
If loastType = ['positive'] then only the example group positive suite will be run.
If loastType = ['spectral', 'positive'] then both Spectral and example group positive suites will be run.

File: src/suites/positive/positive-suite.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (christineyost) [reply src/suites/positive/positive-suite.ts:14]: You could make it required on SuiteConfig instead of optional. Where it's built, `SecurityValuesFactory.buildFromSecuritySchemes()` will always return a `SecurityValues` object. It will just be an empty object if there are none

File: src/commands/suites.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/commands/suites.ts:44]: Test name should follow the pattern  "API <apiName> v<version> env <targetEnv>"
The commands pass "test" instead.

File: src/suites/spectral/spectral-suite.ts [hunk 18]
```diff
@@ -0,0 +1,53 @@
+import { OperationResult } from '../../validation';
+import Suite, { SuiteConfig } from '../suite';
+import SpectralValidator from './validation/spectral-validator';
+
+export default class SpectralSuite extends Suite {
+  public static suiteId = 'spectral';
+  public static label = '(Spectral)';
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/spectral/spectral-suite.ts]: That's the best part of that approach.  We can change both the suiteId and label with ease.

File: src/suites/positive/positive-suite.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
+import { OperationResult } from '../../validation';
+import { OperationExampleFactory } from '../../oas-parsing/operation-example';
+import { ValidationConductor } from './conductors';
+import Suite from '../suite';
+
+export default class PositiveSuite extends Suite {
+  public static suiteId = 'positive';
+  public static label = '(Example Group: Positive)';
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/positive/positive-suite.ts]: works for me

File: src/suites/suite.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+import { SecurityValues } from 'swagger-client';
+import OASSchema from '../oas-parsing/schema';
+
+export interface SuiteConfig {
+  options: TestOptions;
+  schema: OASSchema;
+  securityValues?: SecurityValues;
+}
+
+// New test suites should be added under the '/suites/<suite name>' folder and defined in the SuiteFactory
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/suite.ts:12]: The naming is trying to follow terminology of other unit test conventions encountered in the past.  
- Test is just a single check against given criteria
- Test suite is a collection of related tests.  I've grouped the tests into the two test suites based on what they are attempting to test against. Example group req/resp checking vs linting.  In our case this is extra handy since different tooling drives the two test suites.
- Suite set being a combination of test suites.  Not all APIs should be getting the same test suites at the end of the day.
- Batch being a group of samples(API/ver.#) that a set of test are applied to.

The only thing that changed for the commands was they no longer run a the positive suite but instead they run a 'suite set'.

[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/suite.ts:12]: If wanted, we could rename the commands to "suitesset" vs a "suitesset-batch".  I personally find it more confusing.

File: src/suites/spectral/spectral-suite.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (christineyost) [reply src/suites/spectral/spectral-suite.ts:47]: I think this label was also public but it's protected now

File: src/loast.ts [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    this.relevantSecuritySchemes =
+      await this.suiteConfig.schema.getRelevantSecuritySchemes();
+
+    const securityValues = SecurityValuesFactory.buildFromSecuritySchemes(
+      this.relevantSecuritySchemes,
+      this.options.apiKey,
+      this.options.token,
+    );
+
+    this.suiteConfig.securityValues = securityValues;
```
Comment (christineyost) [inline src/loast.ts:90]:
```suggestion
    const schema = new OASSchema(oasSchemaOptions);
    
    this.relevantSecuritySchemes = await schema.getRelevantSecuritySchemes();
    
    const securityValues = SecurityValuesFactory.buildFromSecuritySchemes(
      this.relevantSecuritySchemes,
      this.options.apiKey,
      this.options.token,
    );

    this.suiteConfig = {
      options: this.options,
      schema,
      securityValues
    };
```

File: src/suites/positive/positive-suite.ts [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+export default class PositiveSuite extends Suite {
+  public static suiteId = 'positive';
+  protected static label = '(Example Group: Positive)';
+
+  async conduct(): Promise<OperationResult[]> {
+    this.checkTargetServer();
+
+    if (!this.suiteConfig.securityValues) {
+      throw new Error('Unable to run suite due to missing securityValues');
+    }
```
Comment (christineyost) [inline src/suites/positive/positive-suite.ts:15]:
```suggestion
```

File: src/suites/suite.ts (hunk 6 above)
Comment (christineyost) [inline src/suites/suite.ts:9]:
```suggestion
  securityValues: SecurityValues;
```

File: test/suites/spectral/spectral-suite.test.ts [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+  new Map().set('no-rings-found', {
+    failures: { test: 'mockError' },
+    warnings: {},
+  }),
+);
+
+describe('SpectralSuite', () => {
+  beforeEach(() => {
+    oasSchema = new OASSchema({ spec: {} });
+    suiteConfig = { options: { path: 'pathToTheRing' }, schema: oasSchema };
```
Comment (christineyost) [inline test/suites/spectral/spectral-suite.test.ts:53]:
```suggestion
    suiteConfig = { 
      options: { path: 'pathToTheRing' }, 
      schema: oasSchema,
      securityValues: {} 
    };
```

**Review by christineyost — COMMENTED** [review]
We shouldn't need that check for suiteConfig.securityValues not being defined in PositiveSuite, since `SecurityValues.buildFromSecuritySchemes()` always returns a SecurityValue object, suiteConfig.securityValues will always be defined. It doesn't need to be marked as optional in `SuiteConfig`.

I do think we should work to remove securityValues from the base suite config, but in the meantime this would be a good change to make

File: src/suites/positive/positive-suite.ts (hunk 22 above)
Comment (mheusgit) [reply src/suites/positive/positive-suite.ts:16]: Looks identical

Comment (christineyost) [reply src/suites/positive/positive-suite.ts:16]: I'm suggesting we remove it

Comment (mheusgit) [reply src/suites/positive/positive-suite.ts:16]: Alright

File: src/loast.ts (hunk 21 above)
Comment (mheusgit) [reply src/loast.ts]: Its been updated

**Comment by mheusgit** [comment]
PR should be good to go

**Review by christineyost — APPROVED** [review]
Changes look good

**Review by dsteeve — APPROVED** [review]
yeah, I think we can go ahead with this change and put in the tickets we discussed.
